### PR TITLE
Preserve annotation default values when extracting Java APIs

### PR DESCRIPTION
**Pull request by lptr**
Looks like Java API normalization did not retain annotation default values. This now adds support for all supported default value types, like so:

```java
@interface Ann {
    int[] arr() default {};
    Deprecated child() default @Deprecated(since = "1.5");
    ElementType enum0() default ElementType.TYPE;
    String name() default "name";
    Class<?> type() default Integer.class;
    int value() default 42;
}
```

Fixes https://github.com/gradle/gradle/issues/30285

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by lptr** [comment]
I added what I expected to be a fix, and it does work for primitive, `String` and `Class`. However, it doesn’t seem to work for enums, arrays and other annotations.

https://github.com/gradle/gradle/pull/30277/files#diff-ef8d4907571d7b59424f1ef846a231555d388b91132b755b7509f3462a0da47eR330-R332

If I uncomment any of these lines, I get failures along the lines of:

```text
Signature Parse error: Expected Field Type Signature
	Remaining input: ()[I
java.lang.reflect.GenericSignatureFormatError: Signature Parse error: Expected Field Type Signature
	Remaining input: ()[I
	at java.base/java.lang.Class.createAnnotationData(Class.java:3757)
	at java.base/java.lang.Class.annotationData(Class.java:3746)
	at java.base/java.lang.Class.getAnnotation(Class.java:3651)
	at java.desktop/java.beans.BeanDescriptor.<init>(BeanDescriptor.java:73)
	at java.desktop/java.beans.Introspector.getTargetBeanDescriptor(Introspector.java:1119)
	at java.desktop/java.beans.Introspector.getBeanInfo(Introspector.java:444)
	at java.desktop/java.beans.Introspector.getBeanInfo(Introspector.java:195)
	at org.gradle.internal.tools.api.ApiClassExtractorAnnotationsTest.default values for annotations are retained(ApiClassExtractorAnnotationsTest.groovy:346)
```

I'm not sure what this refers to.

File: build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorAnnotationsTest.groovy [hunk 1]
```diff
@@ -310,6 +313,75 @@ class ApiClassExtractorAnnotationsTest extends ApiClassExtractorTestSupport {
...
+            Ann: '''
+                import java.lang.annotation.ElementType;
+                import java.lang.annotation.Retention;
+                import java.lang.annotation.RetentionPolicy;
+                import java.lang.annotation.Target;
+
+                @Retention(RetentionPolicy.RUNTIME)
+                @Target({ElementType.TYPE})
+                @interface Ann {
+                    int[] arr() default {};
```
Comment (wolfs) [inline build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorAnnotationsTest.groovy:331]: 🤔 Does it make a difference whether the array has elements in the test?

**Review by wolfs — APPROVED** [review]
Looks good! I'll try it out when there is a nightly of the Gradle API jar.

File: build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorAnnotationsTest.groovy (hunk 1 above)
Comment (lptr) [reply build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorAnnotationsTest.groovy:331]: Only empty arrays are allowed as defaults for arrays parameters.

**Pull request merged**